server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -520,30 +521,62 @@ public void close() throws IOException {
         private final BigArrays bigArrays;
 
         boolean fullStateWritten = false;
+        private final AtomicBoolean closed = new AtomicBoolean();
 
         private Writer(List<MetaDataIndexWriter> metaDataIndexWriters, String nodeId, BigArrays bigArrays) {
             this.metaDataIndexWriters = metaDataIndexWriters;
             this.nodeId = nodeId;
             this.bigArrays = bigArrays;
         }
 
+        private void ensureOpen() {
```

Comment:
I'm surprised we need to check this, given that the caller has already checked we're open in `reloadWriterIfNecessary()`. Can you clarify?

Reply:
This was more separation of concerns. I did not want `Writer` here to rely on correct usage by the caller, but enforce it as well.

Reply:
Maybe we should be asserting that it's open instead?

Reply:
Now I remembered. We can't make this an assertion as it breaks due to concurrency between close() and the other methods.